core/src/mender-client.c

```diff
@@ -363,26 +338,28 @@ mender_client_init(mender_client_config_t *config, mender_client_callbacks_t *ca
         return ret;
     }
 
-    /* Register the rootfs-image update module */
-    if (NULL == (rootfs_image_umod = malloc(sizeof(mender_update_module_t)))) {
-        mender_log_error("Unable to allocate memory for the 'rootfs-image' update module");
+#ifdef CONFIG_MENDER_ZEPHYR_IMAGE_UPDATE_MODULE
+    /* Register the zephyr-image update module */
+    if (NULL == (zephyr_image_umod = calloc(1, sizeof(mender_update_module_t)))) {
+        mender_log_error("Unable to allocate memory for the 'zephyr-image' update module");
         ret = MENDER_FAIL;
         goto END;
     }
-    rootfs_image_umod->callbacks[MENDER_UPDATE_STATE_DOWNLOAD] = &rootfs_img_download_artifact_flash_callback;
-    rootfs_image_umod->callbacks[MENDER_UPDATE_STATE_INSTALL]  = &rootfs_img_ensure_pending_image;
-    rootfs_image_umod->callbacks[MENDER_UPDATE_STATE_REBOOT]   = &rootfs_img_reboot_callback;
-    rootfs_image_umod->callbacks[MENDER_UPDATE_STATE_FAILURE]  = &rootfs_img_ensure_abort_deployment;
-    rootfs_image_umod->artifact_type                           = "rootfs-image";
-    rootfs_image_umod->requires_reboot                         = true;
-    rootfs_image_umod->supports_rollback                       = false; /* TODO: support rollback */
+    zephyr_image_umod->callbacks[MENDER_UPDATE_STATE_DOWNLOAD] = &mender_zephyr_image_download_artifact_flash_callback;
+    zephyr_image_umod->callbacks[MENDER_UPDATE_STATE_INSTALL]  = &mender_zephyr_image_set_pending_image;
+    zephyr_image_umod->callbacks[MENDER_UPDATE_STATE_REBOOT]   = &mender_zephyr_image_reboot_callback;
+    zephyr_image_umod->callbacks[MENDER_UPDATE_STATE_FAILURE]  = &mender_zephyr_image_abort_deployment;
+    zephyr_image_umod->artifact_type                           = "zephyr-image";
+    zephyr_image_umod->requires_reboot                         = true;
+    zephyr_image_umod->supports_rollback                       = false; /* TODO: support rollback */
```

Comment:
If all this initialization is happening in `mender-client.c`, it is not clear to me how will I register my custom update module. We should have an entry point in the API for it.

Reply:
It's there -- [`mender_client_register_update_module()`](https://github.com/mendersoftware/mender-mcu/blob/feature-update_modules/include/mender-client.h#L81). You have to initialize the update module yourself, of course.

Reply:
Yet we are still assigning all the properties here, like the artifact_type, the callbacks, etc.

Let me comment this from a different point of view: I would expect `mender-client.c` to know nothing about `zephyr-image` updates, so no `mender_update_module_t *zephyr_image_umod` declared here nor any code under `CONFIG_MENDER_ZEPHYR_IMAGE_UPDATE_MODULE`. Isn't that possible? And does it make sense to you?

Reply:
oh, right, so you mean it shouldn't be in `mender-client.c` but rather in the code that initializes the client. That makes sense, but having it here means a user code that wants to use the `zephyr-image` type doesn't have to worry about update modules at all. I don't know if that is desirable or not. I can definitely at least move this code into a function in `mender-zephyr-image-update-module.c/h` and then it can be called from one place or another.

Reply:
> oh, right, so you mean it shouldn't be in `mender-client.c` but rather in the code that initializes the client. 

Yes! Sorry

> That makes sense, but having it here means a user code that wants to use the `zephyr-image` type doesn't have to worry about update modules at all. I don't know if that is desirable or not.

The thing is, that I would treat "zephyr-image" Update Module any different than the rest. We develop it ourselves and can have an easier KConfig/CMake integration than the rest, but code wise it should be clear that it is just one update module and showcase everything that you need to do in order to create your own.

> I can definitely at least move this code into a function in `mender-zephyr-image-update-module.c/h` and then it can be called from one place or another.

That sounds good to me :+1: